src/code/z_rumble.c

```diff
@@ -0,0 +1,142 @@
+        strength = sourceStrength - (dist * 255) / 1000;
+
+        if (strength > 0) {
+            // Note: sRumbleMgr is a shared resource between the graph and padmgr threads, no locking is done
+            // to ensure that the entire request is written before it is possibly used.
+            sRumbleMgr.overrideStrength = strength;
+            sRumbleMgr.overrideDuration = duration;
+            sRumbleMgr.overrideDecreaseRate = decreaseRate;
+        }
+    }
+}
+
+/**
+ * Submits a request to the rumble manager with the properties given in the arguments. If there is no free request slot
+ * the request is silently dropped.
+ *
+ * @param distSq
+ *  Squared distance, usually taken to be from an apparent source to the player in world coordinates.
+ * @param sourceStrength
+ *  The strength of the rumble at 0 distance from the source.
+ *  The rumble source strength decreases linearly with distance, a distance of 0 results in the full source strength
+ *  while a distance of 1000 or greater is discarded. A source strength of 0 is discarded. A minimum source strength of
+ *  1 drops to 0 at 3 units of distance from the source. A maximum source strength of 255 drops to 0 at 1000 units of
+ *  distance from the source.
+ *  Note that the strength does not decrease by changing the distance to the source once the request has been submitted.
+ * @param duration
+ *  The duration for which the rumble will sustain full strength. It is measured in Vertical Retraces rather than game
+ *  frames. There are ~60 Retraces/sec on NTSC and 50 Retraces/sec on PAL.
+ * @param decreaseRate
+ *  The amount by which to lower the strength every Vertical Retrace once duration has hit 0.
+ */
+void Rumble_Request(f32 distSq, u8 sourceStrength, u8 duration, u8 decreaseRate) {
+    s32 dist;
+    s32 strength;
+    s32 i;
+
+    if (distSq > SQ(1000)) {
+        dist = 1000;
+    } else {
+        dist = sqrtf(distSq);
+    }
+
+    if (dist < 1000 && sourceStrength != 0 && decreaseRate != 0) {
+        // Decrease the strength linearly with distance
+        strength = sourceStrength - (dist * 255) / 1000;
+
+        for (i = 0; i < RUMBLE_MAX_REQUESTS; i++) {
+            // Search for an empty slot
+            if (sRumbleMgr.reqStrengths[i] == 0) {
+                if (strength > 0) {
+                    // Note: sRumbleMgr is a shared resource between the graph and padmgr threads, no locking is done
+                    // to ensure that the entire request is written before it is possibly used.
+                    sRumbleMgr.reqStrengths[i] = strength;
+                    sRumbleMgr.reqDurations[i] = duration;
+                    sRumbleMgr.reqDecreaseRates[i] = decreaseRate;
+                }
+                break;
+            }
+        }
+    }
+}
+
+void Rumble_Init(void) {
+    RumbleMgr_Init(&sRumbleMgr);
+
+    gPadMgr.retraceCallback = Rumble_Update;
+    gPadMgr.retraceCallbackValue = 0;
+
+    if (1) {}
```

Comment:
Given that this was made a separate function in MM, I wonder if a macro would aid matching?